File: tests/debug_smoke.lua
-- Driven by debug_smoke.py against a real LLDB adapter and temporary C++ program.
vim.o.columns = 180
vim.o.lines = 50
assert(not package.loaded.dap, 'debugger must not load during ordinary startup')
require('lazy').load { plugins = { 'nvim-dap' } }
local dap = require 'dap'
local result = { stops = {} }
local done = false
local function finish()
  if done then
    return
  end
  done = true
  vim.fn.writefile({ vim.json.encode(result) }, vim.env.NVIM_DAP_RESULT)
  vim.schedule(function()
    vim.cmd 'qa!'
  end)
end

dap.listeners.after.event_stopped['smoke'] = function(session, body)
  result.stops[#result.stops + 1] = body.reason
  -- Let the normal DAP handler select the stopped thread before stepping it.
  vim.defer_fn(function()
    session:request('stackTrace', { threadId = body.threadId }, function(err, response)
      if err then
        result.error = err
        finish()
        return
      end
      local frame = response.stackFrames[1]
      if #result.stops == 1 then
        result.frame = { name = frame.name, line = frame.line, path = frame.source.path }
        session:request('evaluate', { expression = 'a + b', frameId = frame.id, context = 'watch' }, function(evalerr, value)
          result.evaluation = value and value.result
          result.evaluate_error = evalerr
          dap.step_over()
        end)
      else
        result.step_line = frame.line
        dap.continue()
      end
    end)
  end, 200)
end
dap.listeners.after.event_exited['smoke'] = function(_, body)
  result.exitCode = body.exitCode
  finish()
end
dap.listeners.after.event_terminated['smoke'] = function()
  vim.defer_fn(finish, 200)
end
vim.defer_fn(function()
  vim.api.nvim_win_set_cursor(0, { 2, 0 })
  dap.toggle_breakpoint()
  local configs = dap.providers.configs['dap.launch.json']()
  assert(#configs == 1, 'project launch.json must be discovered automatically')
  dap.run(configs[1])
end, 300)
vim.defer_fn(function()
  if not done then
    result.timeout = true
    dap.terminate()
    finish()
  end
end, 15000)

File: tests/debug_smoke.py
"""Exercise native launch.json discovery, breakpoint, evaluation, stepping and exit."""
import json
import os
from pathlib import Path
import shutil
import subprocess
import tempfile


repo = Path(__file__).resolve().parents[1]
with tempfile.TemporaryDirectory(prefix="nvim-debug-test-") as directory:
    # macOS /var aliases /private/var; debug info must match Neovim's canonical path.
    root = Path(directory).resolve()
    source = root / "main.cpp"
    source.write_text("""int add(int a, int b) {
    int result = a + b;
    return result;
}
int main() {
    int value = add(19, 23);
    return value == 42 ? 0 : 1;
}
""")
    compiler = shutil.which("clang++") or shutil.which("c++")
    assert compiler, "Install a C++ compiler to run this test"
    subprocess.run([compiler, "-g", "-O0", str(source), "-o", str(root / "app")],
                   cwd=root, check=True)
    (root / ".vscode").mkdir()
    (root / ".vscode/launch.json").write_text(json.dumps({
        "version": "0.2.0",
        "configurations": [{"name": "Native smoke", "type": "lldb", "request": "launch",
                            "program": "${workspaceFolder}/app", "cwd": "${workspaceFolder}",
                            "stopOnEntry": False}],
    }))
    output = root / "result.json"
    proc = subprocess.run(
        ["nvim", "--headless", "-n", "-R", "-i", "NONE", str(source),
         "+luafile " + str(repo / "tests/debug_smoke.lua")],
        cwd=root, env={**os.environ, "NVIM_DAP_RESULT": str(output)},
        capture_output=True, text=True, timeout=25,
    )
    assert proc.returncode == 0, proc.stderr
    result = json.loads(output.read_text())
    assert result.get("stops") == ["breakpoint", "step"], result
    assert result["frame"]["path"] == str(source) and result["frame"]["line"] == 2, result
    assert result.get("evaluation") == "42" and result.get("step_line") == 3, result
    assert result.get("exitCode") == 0 and not result.get("timeout"), result
    assert not result.get("error") and not result.get("evaluate_error"), result
    print("Native debugger: lazy startup, project profile, breakpoint, evaluation, step and clean exit passed")
